_tut/docs/3x/tail/iterant.md

```diff
@@ -0,0 +1,267 @@
+---
+layout: docs3x
+title: Iterant
+type_api: monix.tail.Iterant
+type_source: monix-tail/shared/src/main/scala/monix/tail/Iterant.scala
+description: |
+    Pull-based, compositional streaming data type that can describe lazy, possibly asynchronous events, a purely functional iterator.
+    
+tut:
+  scala: 2.12.4
+  binaryScala: "2.12"
+  dependencies:
+    - io.monix::monix:version3x
+---
+
+## Introduction
+
+The `Iterant` is a data type that describes lazy, possibly asynchronous
+streaming of elements using a pull-based protocol.
+
+It is similar somewhat in spirit to Scala's own
+`collection.immutable.Stream` and with Java's `Iterable`, except
+that it is more composable and more flexible due to evaluation being
+controlled by an `F[_]` data type that you have to supply
+(like [Task](../eval/task.html), [Coeval](../eval/coeval.html)
+or [IO](https://typelevel.org/cats-effect/datatypes/io.html)),
+which will control the evaluation. In other words,
```

Comment:
There's no eager/strict version with `Iterant`. 

You have to consider for one that `Iterant` needs a stack safe `map` and `flatMap` implementations. And this usually implies laziness. `Id` for example wouldn't cut it, see https://github.com/typelevel/cats-effect/issues/92 for why.

And if `F` would be `Future`, which is still safe in `map` and `flatMap`, then the problem is that `Iterant` doesn't make sense. We're trying to model a pull-based protocol, so we have:

1. a `rest`, that in case of `Future` would signal some triggered side effects, but if we have `rest: Future`, then we have no function in there, so `Future` in this context is useless
2. a `stop` that is `F[Unit]` and if `stop: Future[Unit]`, that's also useless, as there's precisely one value that you can assign to it, that being `Future.successful(())`

You could work of course with `() => Future[A]` thunks, but that is no longer eager, being almost equivalent to `Task`. You could have an eagerly initialized `Iterant` too, where all `rest` chains are already evaluated, but all transformations applied on it would yield a lazy `Iterant`.

Reply:
Speaking of which, I need to make this clear in the document, as I now see that the wording is unclear 🤔

Reply:
It is already in Task, Coeval and IO docs - I feel like repeating it every time is a bit redundant :D

Reply:
I agree a minor rewording here will be sufficient, as you both have pointed out. I see now that the section below on Parametric Polymorphism talks about Coeval vs Task, which is very helpful.

Reply:
@kanak please keep the feedback coming. I'm very interested to know if the document is clear for people that aren't very familiar with such data types.

But it's a work in progress, so you can defer any feedback for when I'll finish it if that makes it easier for you.